src/set_theory/ordinal/natural_ops.lean

```diff
@@ -0,0 +1,366 @@
+⟨λ a b c h, by { by_contra' h', exact h.not_le (add_le_add_right h' a) }⟩
+
+instance add_contravariant_class_le :
+  contravariant_class nat_ordinal.{u} nat_ordinal.{u} (+) (≤) :=
+⟨λ a b c h, by { by_contra' h', exact h.not_lt (add_lt_add_left h' a) }⟩
+
+instance swap_add_contravariant_class_le :
+  contravariant_class nat_ordinal.{u} nat_ordinal.{u} (swap (+)) (≤) :=
+⟨λ a b c h, by { by_contra' h', exact h.not_lt (add_lt_add_right h' a) }⟩
+
+theorem lt_add_iff {a b c : nat_ordinal} :
+  a < b + c ↔ (∃ b' < b, a ≤ b' + c) ∨ ∃ c' < c, a ≤ b + c' :=
+by { rw add_def, simpa [blsub, lt_blsub_iff, lt_to_nat_ordinal_iff] }
+
+theorem blsub_add_of_mono {a b : nat_ordinal.{u}} {f : Π c < (a + b), nat_ordinal.{max u v}}
+  (hf : ∀ {i j} hi hj, i ≤ j → f i hi ≤ f j hj) : blsub _ f = max
+  (blsub.{u v} a (λ a' ha', f (a' + b) $ add_lt_add_right ha' b))
+  (blsub.{u v} b (λ b' hb', f (a + b') $ add_lt_add_left hb' a)) :=
+begin
+  apply le_antisymm (blsub_le_iff.2 (λ i hi, _)) (max_le _ _),
+  { rcases lt_add_iff.1 hi with ⟨a', ha', hi'⟩ | ⟨b', hb', hi'⟩,
+    { exact lt_max_of_lt_left ((hf hi (add_lt_add_right ha' b) hi').trans_lt (lt_blsub _ a' _)) },
+    { exact lt_max_of_lt_right ((hf hi (add_lt_add_left hb' a) hi').trans_lt (lt_blsub _ b' _)) } },
+  all_goals
+  { apply blsub_le_of_brange_subset.{u u v},
+    rintro c ⟨d, hd, rfl⟩,
+    apply mem_brange_self }
+end
+
+private theorem add_assoc' : ∀ a b c : nat_ordinal, a + b + c = a + (b + c)
+| a b c := begin
+  nth_rewrite 2 add_def,
+  rw [add_def, blsub_add_of_mono, blsub_add_of_mono, max_assoc],
+  { congr;
+    ext d hd;
+    apply add_assoc' },
+  { exact λ i j _ _ h, add_le_add_left h a },
+  { exact λ i j _ _ h, add_le_add_right h c }
+end
+using_well_founded { dec_tac := `[solve_by_elim [psigma.lex.left, psigma.lex.right]] }
+
+instance : add_semigroup nat_ordinal :=
+{ add := (+),
+  add_assoc := add_assoc' }
+
+private theorem add_comm' : ∀ a b : nat_ordinal, a + b = b + a
+| a b := begin
+  rw [add_def, add_def, max_comm],
+  congr;
+  ext a' ha';
+  apply add_comm'
+end
+using_well_founded { dec_tac := `[solve_by_elim [psigma.lex.left, psigma.lex.right]] }
+
+instance : add_comm_semigroup nat_ordinal :=
+{ add_comm := add_comm',
+  ..nat_ordinal.add_semigroup }
+
+private theorem add_zero' (a : nat_ordinal) : a + 0 = a :=
+begin
+  induction a using nat_ordinal.induction with a IH,
+  rw add_def,
+  simp only [blsub, of_nat_ordinal_zero, blsub_zero, to_nat_ordinal_zero, max_zero_right],
+  rw ←eq_of_nat_ordinal_iff,
+  convert blsub_id (of_nat_ordinal a),
+  ext b hb,
+  exact IH _ hb
+end
+
+instance : add_cancel_comm_monoid nat_ordinal :=
+{ zero := 0,
+  add_zero := add_zero',
+  zero_add := λ a, by rw [add_comm, add_zero'],
+  add_left_cancel := λ a b c, add_left_cancel'',
+  ..nat_ordinal.add_comm_semigroup }
+
+/-- The successor on `nat_ordinal` is still `a + 1`. -/
+theorem add_one (a : nat_ordinal) : a + 1 = to_nat_ordinal (succ (of_nat_ordinal a)) :=
```

Comment:
This lemma might be more useful as
```suggestion
theorem add_one (a : ordinal) : to_nat_ordinal (succ a) = to_nat_ordinal a + 1 :=
```

Reply:
I've chosen to make the lemmas in the `nat_ordinal` namespace primarily about `nat_ordinal`. By contrast, the "boilerplate" lemmas on the `ordinal` namespace are written in terms of `ordinal`.

Reply:
I don't think that's consistent with how we handle these things with `multiplicative`, where the rule would be "lemmas about `nat_ordinal.to_nat_ordinal` should be named  `nat_ordinal.to_nat_ordinal_foo`"

Reply:
The other version that is worth having is
```suggestion
theorem of_nat_ordinal_add_one (a : nat_ordinal) : of_nat_ordinal (a + 1) = succ (of_nat_ordinal a) :=
```

Reply:
I should mention, once #14243 passes and we can transfer the `succ_order` instance to `nat_ordinal`, the original `add_one` would just be saying `a + 1 = succ a`.

Reply:
I rewrote this as `a ♯ 1 = a + 1`. Once I'm able to give a `succ_order` instance on `nat_ordinal`, I'll be able to prove `a + 1 = succ a` by def-eq.